ably/hosts.go

```diff
@@ -0,0 +1,192 @@
+package ably
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ably/ably-go/ably/internal/ablyutil"
+)
+
+// RSC15
+type restHosts struct {
+	primaryFallbackHost string
+	opts                *clientOptions
+	cache               *hostCache
+	visitedHosts        ablyutil.HashSet
+	sync.Mutex
+}
+
+func newRestHosts(opts *clientOptions) *restHosts {
+	return &restHosts{
+		opts: opts,
+		cache: &hostCache{
+			duration: opts.fallbackRetryTimeout(),
+		},
+		visitedHosts: ablyutil.NewHashSet(),
+	}
+}
+
+func (restHosts *restHosts) getPrimaryHost() string {
+	return restHosts.opts.getPrimaryRestHost()
+}
+
+func (restHosts *restHosts) nextFallbackHost() string {
+	restHosts.Lock()
+	defer restHosts.Unlock()
+
+	getNonVisitedHost := func() string {
+		visitedHosts := restHosts.visitedHosts
+		cachedHost := restHosts.cache.get()
+		if !ablyutil.Empty(cachedHost) && !visitedHosts.Has(cachedHost) {
+			return cachedHost
+		}
+		primaryHost := restHosts.getPrimaryHost()
```

Comment:
It can return `primary` host, if `fallback` is cached and it fails next time. So, for all subsequent retries, `primary` host is being treated as `fallback` host.

Reply:
That's really not very clear. We have a primary endpoint, and a set of fallback endpoints, let's please not refer to the primary endpoint as a fallback endpoint, since it isn't a fallback endpoint.

Reply:
What name would you recommend? In the context of retry loop, `nextFallback` makes perfect sense, since we will be trying out fallbacks, doesn't matter if it's a primary host or a fallback host. It's a part of fallback loop. This particular case happens because of cache, this will not happen in case of realtimeHosts, since preferred host is always primary host

Reply:
Also, I am sure any other name would just be more confusing

Reply:
If you read `https://docs.ably.io/client-lib-development-guide/features/#RSC15f` carefully, it states to use primaryFallbackHost if `cachedHost` expires. Implementation is as per spec.